vpr/src/pack/re_cluster.cpp

```diff
@@ -28,44 +41,137 @@ bool move_atom_to_new_cluster(const AtomBlockId& atom_id,
     if (is_created && !during_packing) {
-        fix_clustered_netlist(atom_id, old_clb, new_clb);
+        fix_clustered_netlist(molecule, molecule_size, old_clb, new_clb);
+    }
+
+    for (auto atom_pb_backup : mol_pb_backup) {
+        cleanup_pb(atom_pb_backup);
+        delete atom_pb_backup;
     }
 
     return (is_created);
 }
+
+bool move_mol_to_existing_cluster(t_pack_molecule* molecule,
+                                  const ClusterBlockId& new_clb,
+                                  bool during_packing,
+                                  t_clustering_data& clustering_data) {
+    //define required contexts
+    auto& cluster_ctx = g_vpr_ctx.clustering();
+    auto& atom_ctx = g_vpr_ctx.mutable_atom();
+
+    //define local variables
+    bool is_removed, is_added;
+    AtomBlockId root_atom_id = molecule->atom_block_ids[molecule->root];
+    int molecule_size = get_array_size_of_molecule(molecule);
+    t_lb_router_data* old_router_data = nullptr;
+    std::vector<AtomBlockId> new_clb_atoms = cluster_to_atoms(new_clb);
+
+    //Check that the old and new clusters are the same type
+    ClusterBlockId old_clb = atom_to_cluster(root_atom_id);
+    if (cluster_ctx.clb_nlist.block_type(old_clb) != cluster_ctx.clb_nlist.block_type(new_clb)) {
+        VTR_LOG("Atom:%zu move aborted. New and old cluster blocks are not of the same type",
+                root_atom_id);
+        return false;
+    }
+
+    //Check that the old and new clusters are the mode
+    if (cluster_ctx.clb_nlist.block_pb(old_clb)->mode != cluster_ctx.clb_nlist.block_pb(new_clb)->mode) {
+        VTR_LOG("Atom:%zu move aborted. New and old cluster blocks are not of the same mode",
+                root_atom_id);
+        return false;
+    }
+
+    //Backup the original pb of the atom before the move
+    std::vector<t_pb*> mol_pb_backup;
+    for (int i_atom = 0; i_atom < molecule_size; i_atom++) {
+        if (molecule->atom_block_ids[i_atom]) {
+            t_pb* atom_pb_backup = new t_pb;
+            atom_pb_backup->pb_deep_copy(atom_ctx.lookup.atom_pb(molecule->atom_block_ids[i_atom]));
+            mol_pb_backup.push_back(atom_pb_backup);
+        }
+    }
+
+    //remove the molecule from its current cluster and check the cluster legality
+    is_removed = remove_mol_from_cluster(molecule, molecule_size, old_clb, old_router_data);
+    if (!is_removed) {
+        VTR_LOG("Atom: %zu move failed. Can't remove it from the old cluster\n", root_atom_id);
+        return false;
+    }
+
+    //Add the atom to the new cluster
+    is_added = pack_mol_in_existing_cluster(molecule, new_clb, new_clb_atoms, during_packing, clustering_data);
+
+    //Commit or revert the move
+    if (is_added) {
+        commit_mol_move(old_clb, new_clb, mol_pb_backup, old_router_data, clustering_data, during_packing, false);
+        VTR_LOG("Atom:%zu is moved to a new cluster\n", molecule->atom_block_ids[molecule->root]);
+    } else {
+        int atom_idx = 0;
```

Comment:
Helper function for this else block (seem to have same code twice).
Might be able to make a helper function for the whole if statement too (seems you have that code twice as well).